src/DataSourceProvider.vue

```diff
@@ -1,19 +1,212 @@
+      </div>
+      <DataSourceSelector
+        :currentAppDataSources="appDataSources" :otherDataSources="allDataSources"
+        :selectedDataSource="selectedDataSource"
+        :changeDataSource="changeDataSource"
+        :showAll="showAll"
+        @onDataSourceSelect="onDataSourceSelect"
+        @onDataSourceCreate="onDataSourceCreate"
+        @onShowAll="showAllDataSources"
+        @onDataSourceChange="onDataSourceChange"
+      >
+      </DataSourceSelector>
+      <SecurityNotifier v-show="selectedDataSource"
+        :securityEnabled="hasAccessRules()"
+        @addDefaultSecurity="addDefaultSecurity"
+      >
+      </SecurityNotifier>
+    </div>
+  </section>
+
 </template>
 
 <script>
-import FirstComponent from './components/FirstComponent';
+import DataSourceSelector from './components/DataSourceSelector';
+import SecurityNotifier from './components/SecurityNotifier';
+import { getDataSources, getDataSource, createDataSource, updateDataSourceSecurityRules } from './services/dataSource';
 
 export default {
   data() {
-    return {};
+    return {
+      appDataSources: [],
+      allDataSources: [],
+      copyOfAllDataSources: [],
+      isLoading: true,
+      hasError: false,
+      errorMessage: '',
+      widgetData: {},
+      selectedDataSource: null,
+      changeDataSource: false,
+      showAll: false
+    };
+  },
+  methods: {
+    onDataSourceChange: function() {
+      this.changeDataSource = !this.changeDataSource;
+    },
+    onDataSourceSelect: function(dataSource) {
+      this.selectedDataSource = dataSource;
+    },
+    hasAccessRules: function() {
+      if (!this.selectedDataSource) {
+        return false;
+      }
+
+      if (this.selectedDataSource.accessRules === null || !this.selectedDataSource.accessRules.length) {
+        return false;
+      }
+
+      return true;
+    },
+    addDefaultSecurity: function() {
+      this.isLoading = true;
+      this.selectedDataSource.accessRules = this.widgetData.accessRules;
+
+      updateDataSourceSecurityRules(this.selectedDataSource.id, this.selectedDataSource.accessRules)
+        .then(() => {
+          Fliplet.Modal.alert({
+            message: 'Your changes have been applied to all affected apps.'
+          });
+        })
+        .catch(err => {
+          this.hasError = true;
+          this.errorMessage = Fliplet.parseError(err);
+        })
+        .finally(() => {
+          this.isLoading = false;
+        });
+    },
+    showAllDataSources: function(isChecked) {
+      this.isLoading = true;
+      this.showAll = isChecked;
+
+      if (isChecked) {
+        if (!this.copyOfAllDataSources.length) {
+          this.loadDataSources();
+          return;
+        }
+
+        this.allDataSources = [...this.copyOfAllDataSources];
+      } else {
+        this.copyOfAllDataSources = [...this.allDataSources];
+        this.allDataSources = [];
+      }
+
+      // Give VUE time to reset templates
+      setTimeout(() => {
```

Comment:
Does something like this work? Vue.JS has a method for waiting for the next tick instead of using an arbitrary timeout.

```suggestion
      this.$nextTick(() => {
```

Reply:
@tonytlwu Yes, it works great.

Reply:
@tonytlwu added this change